### PR TITLE
Interdiff mode: corrected comparison of overlapped hunks;

### PR DESCRIPTION
**Pull request by sofiia-tesliuk**
In previous version haven't been considered the situation, when more than 2 hunks are overlapped.

### CODE REVIEW

File: interdiff.go [hunk 1]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
+		StartPosition: -1,
+		Body:          []byte{0},
+	}
+
+	firstOldHunk, firstNewHunk := oldHunks[0], newHunks[0]
+	lastOldHunk, lastNewHunk := oldHunks[len(oldHunks)-1], newHunks[len(newHunks)-1]
+
+	// Calculate StartLine for origin and new in result
+	// Started with old hunk
+	if firstOldHunk.OrigStartLine < firstNewHunk.OrigStartLine {
```
Comment (schachmat) [inline interdiff.go:268]:
```suggestion
	if firstOldHunk.OrigStartLine < firstNewHunk.OrigStartLine {
		// Started with old hunk
```

File: interdiff.go [hunk 2]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
+		currentOrgI = firstOldHunk.OrigStartLine
+		resultHunk.OrigStartLine = firstOldHunk.NewStartLine
+		resultHunk.NewStartLine = firstNewHunk.NewStartLine -
```
Comment (schachmat) [inline interdiff.go:271]: I'm having trouble understanding this computation. Can you explain this a bit in the `// Calculate StartLine for origin and new in result` comment above?

File: interdiff.go [hunk 3]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
+			firstNewHunk.OrigStartLine + currentOrgI
+	} else {
+		// Started with new hunk
+		currentOrgI = firstNewHunk.OrigStartLine
+		resultHunk.OrigStartLine = firstOldHunk.NewStartLine -
+			firstOldHunk.OrigStartLine + currentOrgI
```
Comment (schachmat) [inline interdiff.go:277]:
```suggestion
		resultHunk.NewStartLine = currentOrgI +
			firstNewHunk.NewStartLine - firstNewHunk.OrigStartLine
	} else {
		// Started with new hunk
		currentOrgI = firstNewHunk.OrigStartLine
		resultHunk.OrigStartLine = currentOrgI +
			firstOldHunk.NewStartLine - firstOldHunk.OrigStartLine
```

File: interdiff.go [hunk 4]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
+		resultHunk.NewStartLine = firstNewHunk.NewStartLine
+	}
+
+	// Calculate NumberLines for origin and new in result
+	// Finished with old hunk
+	if lastOldHunk.OrigStartLine+lastOldHunk.OrigLines > lastNewHunk.OrigStartLine+lastNewHunk.OrigLines {
```
Comment (schachmat) [inline interdiff.go:283]:
```suggestion
	if lastOldHunk.OrigStartLine+lastOldHunk.OrigLines > lastNewHunk.OrigStartLine+lastNewHunk.OrigLines {
		// Finished with old hunk
```

File: interdiff.go [hunk 5]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
+	} else {
+		// Finished with new hunk
+		resultHunk.OrigLines = lastOldHunk.NewStartLine + lastOldHunk.NewLines +
+			lastNewHunk.OrigStartLine + lastNewHunk.OrigLines -
+			lastOldHunk.OrigStartLine - lastOldHunk.OrigLines -
+			resultHunk.OrigStartLine
+		resultHunk.NewLines = lastNewHunk.NewStartLine + lastNewHunk.NewLines - resultHunk.NewStartLine
+	}
+
+	// TODO: Check those values
```
Comment (schachmat) [inline interdiff.go:298]: How do you plan to check them? Is this something for a test case?

File: interdiff.go [hunk 6]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
+	resultHunk := &diff.Hunk{OrigStartLine: -1,
+		OrigLines:       -1,
+		OrigNoNewlineAt: -1,
+		NewStartLine:    -1,
+		NewLines:        -1,
+		// TODO: Concatenate sections
+		Section:       "",
+		StartPosition: -1,
+		Body:          []byte{0},
+	}
```
Comment (schachmat) [inline interdiff.go:261]:
```suggestion
	resultHunk := &diff.Hunk{
		// TODO: Concatenate sections
		Section:       "",
		Body:          []byte{0},
	}
```

Members that you set below don't need to be initialized.

File: interdiff.go [hunk 7]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
+					(oldFileDiff.Hunks[i].OrigStartLine < newFileDiff.Hunks[j-1].OrigStartLine+newFileDiff.Hunks[j-1].OrigLines):
+					oldHunks = append(oldHunks, oldFileDiff.Hunks[i])
+					i++
+				// Starting line of new hunk is in old hunk body
+				case (j < len(newFileDiff.Hunks)) && (newFileDiff.Hunks[j].OrigStartLine >= oldFileDiff.Hunks[i-1].OrigStartLine) &&
+					(newFileDiff.Hunks[j].OrigStartLine < oldFileDiff.Hunks[i-1].OrigStartLine+oldFileDiff.Hunks[i-1].OrigLines):
+					newHunks = append(newHunks, newFileDiff.Hunks[j])
+					j++
+				default:
+					findAll = true
```
[the remark this answers is not in the document]
Comment (sofiia-tesliuk) [reply interdiff.go:111]: In this case `break` will break out from switch, but not from surrounding loop.
Marked loop as Loop and used `break Loop` instead.

[the remark this answers is not in the document]
Comment (schachmat) [reply interdiff.go:111]: yes, good catch!

File: interdiff.go [hunk 8]
```diff
@@ -87,40 +84,222 @@ func compareFileDiff(oldFileDiff, newFileDiff *diff.FileDiff) *diff.FileDiff {
...
 	return &resultFileDiff
 }
 
-func compareHunk(oldHunk, newHunk *diff.Hunk) *diff.Hunk {
-	// diffs are the same
-	if (oldHunk.OrigStartLine == newHunk.OrigStartLine) &&
-		(oldHunk.OrigLines == newHunk.OrigLines) &&
-		(oldHunk.OrigNoNewlineAt == newHunk.OrigNoNewlineAt) &&
-		(bytes.Equal(oldHunk.Body, newHunk.Body)) {
+func compareOverlappedHunks(oldHunks, newHunks []*diff.Hunk) *diff.Hunk {
```
[the remark this answers is not in the document]
Comment (sofiia-tesliuk) [reply interdiff.go:137]: Will be added in another PR.

[the remark this answers is not in the document]
Comment (schachmat) [reply interdiff.go:137]: ack, thx.

File: interdiff.go (hunk 5 above)
Comment (sofiia-tesliuk) [reply interdiff.go:298]: I'm not sure where those values will be used and if they are correct in that way as they are right now (like [StartPosition](https://github.com/sourcegraph/go-diff/blob/f935979a5d70d2f3034edc6de94c77be621ce500/diff/diff.pb.go#L76)).

Comment (schachmat) [reply interdiff.go:298]: Seems this is only used here: https://github.com/sourcegraph/go-diff/blob/f935979a5d70d2f3034edc6de94c77be621ce500/diff/parse.go#L605

I don't think we need to care about this for the result hunk at all.

Not sure about the `OrigNoNewlineAt` one. According to https://github.com/sourcegraph/go-diff/blob/3f415a150aec0685cb81b73cc201e762e075006d/diff/print.go#L104 I'd set it to `0` for now, but we probably need to actually handle it correctly in a followup.

File: interdiff.go [hunk 9]
```diff
@@ -236,70 +284,81 @@ func compareOverlappedHunks(oldHunks, newHunks []*diff.Hunk) *diff.Hunk {
...
 
-	return nil
+	return nil, nil
 }
 
-func configureResultHunk(oldHunks, newHunks []*diff.Hunk) (*diff.Hunk, int32) {
+func configureResultHunk(oldHunks, newHunks []*diff.Hunk) (*diff.Hunk, int32, error) {
 	if (len(oldHunks) == 0) || (len(newHunks) == 0) {
-		return nil, 0
+		return nil, 0, fmt.Errorf("one of the hunks array is empty")
```
Comment (schachmat) [inline interdiff.go:297]:
```suggestion
		return nil, 0, errors.New("one of the hunks array is empty")
```

If you're not using any format strings, it's recommended to use `errors.New()` which just takes a regular string value. You'll also need to add that new import.

File: interdiff.go (hunk 5 above)
Comment (sofiia-tesliuk) [reply interdiff.go:357]: Set it to 0 for now.

**Pull request merged**